### PR TITLE
[GemmaRMSNorm] Fix GemmaRMSNorm not dispatching to fused kernel

### PR DESCRIPTION
**Pull request by mayuyuace**
**Problem**:
GemmaRMSNorm.forward_native computed weight as self.weight.float() + 1.0, producing a float32 tensor. The vllm_c kernel's supports_args check (rms_no_var_size) requires weight.dtype == x.dtype, so with float32 weight and bfloat16 input the check fails and dispatch always falls back to the native PyTorch decomposition (~8 kernel launches per norm).

This is not platform-specific; the issue exists on both CUDA and XPU.

**Fix**:
GemmaRMSNorm.forward_native: remove .float() so weight stays in model dtype (bfloat16).
In this way, `forward_native` will call the fused kernel instead of the native implementation.
The changes in `vllm/ir/ops/layernorm.py` are intended to eliminate the requirement to convert weights to FP32 precision when invoking the native implementation in other scenarios.
The fused kernels already promote to float32 internally for variance accumulation, so numerical correctness is preserved.

### CODE REVIEW

File: vllm/ir/ops/layernorm.py
Comment (zufangzhu) [inline vllm/ir/ops/layernorm.py:15]: x is float32, no need to do cast again

Comment (mayuyuace) [reply vllm/ir/ops/layernorm.py:15]: OK. Removed the data type cast.

**Comment by mayuyuace** [comment]
@yewentao256 Can you help review this PR if you have some free time?

File: vllm/model_executor/layers/layernorm.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def forward_native(
         residual: torch.Tensor | None = None,
     ) -> torch.Tensor | tuple[torch.Tensor, torch.Tensor]:
         """PyTorch-native implementation equivalent to forward()."""
-        weight = self.weight.float() + 1.0
+        weight = self.weight + 1.0
```
Comment (ccrhx4) [inline vllm/model_executor/layers/layernorm.py:157]: This change alters the numeric behavior.

Original: weight.float() + 1
After: float(weight + 1)

They will yield different results if bf16 has a value with a fractional component whose exact adjustment requires more precision than bf16 can hold, or if the addition causes a rounding difference due to bit truncation in the 7-bit mantissa of bfloat16.

#### For example, bf16 = 1.00390625 
- This specific number can just barely be represented in bfloat16
#### Evaluating: (bf16 + 1).to(float)
- First, it does 1.00390625 + 1.0 in bfloat16.
- The exact math is 2.00390625.
- However, bfloat16 at the scale of 2.0 can only count in steps of 0.0078125. It does not have enough bits to hold the .00390625 fraction anymore.
- The hardware rounds it to the nearest representable bf16 value, which is exactly 2.0.
- Converting 2.0 to float gives 2.0.
#### Evaluating: float(bf16) + 1
- First, it converts 1.00390625 to a 32-bit float.
- float32 has plenty of bits, so it holds 1.00390625 perfectly.
- Then it adds 1.0 using float32 precision.
- The result is exactly 2.00390625
#### The Result

- (bf16 + 1).to(float) = 2.0

- float(bf16) + 1 = 2.00390625

Comment (mayuyuace) [reply vllm/model_executor/layers/layernorm.py:157]: Tested with Qwen/Qwen3.5-35B-A3B on B60 and command:
```
lm_eval --model local-completions --tasks gsm8k --model_args "model=$model,base_url=http://localhost:30088/v1/completions,max_gen_toks=1024,num_concurrent=64,timeout=9999999,temperature=0"
```
Results of full gsm8k tasks:

```
Original:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9067|±  |0.0080|
|     |       |strict-match    |     5|exact_match|↑  |0.9098|±  |0.0079|

With this PR:
Qwen/Qwen3.5-35B-A3B
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9196|±  |0.0075|
|     |       |strict-match    |     5|exact_match|↑  |0.9189|±  |0.0075|

```
There is no regression.

Comment (jikunshang) [reply vllm/model_executor/layers/layernorm.py:157]: possible to post a cuda acc data?

Comment (mayuyuace) [reply vllm/model_executor/layers/layernorm.py:157]: > possible to post a cuda acc data?

Qwen/Qwen3.5-35B-A3B on L20: 
```
Original:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9037|±  |0.0081|
|     |       |strict-match    |     5|exact_match|↑  |0.9075|±  |0.0080|

With this PR:
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9067|±  |0.0080|
|     |       |strict-match    |     5|exact_match|↑  |0.9121|±  |0.0078|
```

**Comment by mayuyuace** [comment]
@ccrhx4 

## 1 Quantifying the actual error magnitude

In bfloat16, numbers in the range [1.0, 2.0) are spaced at intervals of 2^(-7) = 0.0078125. So when we compute w + 1 in bf16 (where w is a small trained delta, e.g. 0.05), the result 1.05 gets rounded to the nearest representable bf16 value (1.046875), introducing an absolute error of ~0.003125 — a relative error of roughly 0.3% on the scale factor (w+1). This error is bounded; the worst case is half a bf16 step, ~0.004.

## 2 Why 0.3% scale error does not affect model accuracy

There are two complementary reasons:

First, the fused kernel still promotes to float32 internally for the variance accumulation (i.e., computing sum(x²)). This is the numerically sensitive part of RMSNorm. The bf16 rounding only affects the final element-wise multiply by (w+1), which is a simple per-channel scaling applied to already-normalized values. A 0.3% perturbation on a per-channel scale factor as below show, after values have been normalized to roughly unit variance, produces a negligibly small absolute deviation in the output.

The Gemma RMSNorm output is:

$$\hat{x} = \frac{x}{\text{RMS}(x)} \cdot (w + 1)$$

When `(w+1)` is computed in bf16, it introduces a rounding error $\epsilon$, so the actual output becomes:

$$\hat{x}' = \frac{x}{\text{RMS}(x)} \cdot (w + 1 + \epsilon)$$

The relative error on each output element is:

$$\frac{|\hat{x}' - \hat{x}|}{|\hat{x}|} = \frac{|\epsilon|}{|w + 1|}$$

Since $w$ is a small trained delta initialized from 0, $|w+1| \approx 1$, so the relative error is simply $|\epsilon| \approx 0.003$ (~0.3%).

Second, large language models are inherently robust to small weight perturbations of this magnitude. A bounded bf16 rounding error of ~0.003 on the LayerNorm scale is well within the tolerance of any downstream computation (linear projections, attention softmax, GELU activations), and this is confirmed empirically by the gsm8k results showing no regression across both XPU (B60) and CUDA (L20) hardware.

**Review by yewentao256 — COMMENTED** [review]
Hi @mayuyuace Thanks for the work! I don't think I have enough context to review the PR, you might want to find another reviewer as well

Here are some AI generated comments

```bash
Thanks for investigating this. I still think this changeset needs revision before merging.
First, self.weight + 1.0 performs the addition in BF16/FP16, whereas Gemma’s reference implementation explicitly computes self.weight.float() + 1.0. The GSM8K result does not establish numerical equivalence across all models using GemmaRMSNorm. In a BF16 32×1024 GPU repro, 8,697 output elements changed without a residual and 11,290 changed with a residual, with a maximum absolute difference of 0.03125. The quoted ~0.004 bound applies to the scale-factor rounding error, not to the final output error.
Second, changing the shared IR implementation from x.to(weight.dtype) * weight to x * weight.float() alters the native semantics of every standard RMSNorm, not just Gemma. This is a broader, unrelated behavioral change and also differs from the standard Llama RMSNorm rounding order.
Could we instead extend the fused kernel with a Gemma-style weight_bias=1.0, applying float(weight) + 1.0f inside the kernel? That would preserve the existing semantics while still enabling fused dispatch.
Please also add BF16/FP16 regression tests for both residual and non-residual paths that explicitly verify vllm_c is selected. The existing mixed-dtype test keeps the weight in FP32 and therefore does not exercise the new fused path.
```
